File: examples/php/12-testing/tests/CalculatorTest.php
<?php
declare(strict_types=1);

use PHPUnit\Framework\TestCase;

require_once __DIR__ . '/../src/Calculator.php';

/**
 * Тема 12 – Тестване: Unit тест за Calculator
 *
 * Стартиране:
 *   composer install
 *   vendor/bin/phpunit tests/
 *
 * Основни PHPUnit концепции:
 *   - Test class наследява TestCase
 *   - Всеки test метод е публичен и започва с "test" (или има @test анотация)
 *   - assert* методи проверяват очакваните резултати
 *   - setUp() се изпълнява преди всеки тест
 *   - @dataProvider захранва тест с множество входни данни
 */
class CalculatorTest extends TestCase
{
    private Calculator $calc;

    // ── setUp се изпълнява преди ВСЕКИ тест ──────────────────────────
    protected function setUp(): void
    {
        $this->calc = new Calculator();
    }

    // ── Основни тестове ──────────────────────────────────────────────

    public function testAdd(): void
    {
        $this->assertSame(5.0, $this->calc->add(2, 3));
        $this->assertSame(0.0, $this->calc->add(-1, 1));
    }

    public function testSubtract(): void
    {
        $this->assertSame(1.0, $this->calc->subtract(3, 2));
        $this->assertSame(-5.0, $this->calc->subtract(0, 5));
    }

    public function testMultiply(): void
    {
        $this->assertSame(12.0, $this->calc->multiply(3, 4));
        $this->assertSame(0.0, $this->calc->multiply(5, 0));
        $this->assertSame(-6.0, $this->calc->multiply(-2, 3));
    }

    public function testDivide(): void
    {
        $this->assertSame(2.5, $this->calc->divide(5, 2));
        $this->assertSame(-3.0, $this->calc->divide(-6, 2));
    }

    // ── Тест за изключение ───────────────────────────────────────────

    public function testDivideByZeroThrowsException(): void
    {
        // Очакваме InvalidArgumentException при деление на нула
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Division by zero');

        $this->calc->divide(10, 0);
    }

    // ── Data Provider ────────────────────────────────────────────────
    // Тества се с множество входни стойности от един тест

    /**
     * @dataProvider percentageProvider
     */
    public function testPercentage(float $value, float $pct, float $expected): void
    {
        $this->assertEqualsWithDelta($expected, $this->calc->percentage($value, $pct), 0.001);
    }

    /** @return array<string, array{float, float, float}> */
    public static function percentageProvider(): array
    {
        return [
            '10% от 200'  => [200.0, 10.0, 20.0],
            '50% от 100'  => [100.0, 50.0, 50.0],
            '0% от 500'   => [500.0,  0.0,  0.0],
            '100% от 75'  => [ 75.0,100.0, 75.0],
            '7.5% от 400' => [400.0,  7.5, 30.0],
        ];
    }

    // ── Групиране (annotations) ─────────────────────────────────────

    /**
     * @group arithmetic
     */
    public function testAddIsCommutative(): void
    {
        $a = 7.3;
        $b = 2.1;
        $this->assertSame($this->calc->add($a, $b), $this->calc->add($b, $a));
    }
}

File: examples/java/12-testing/src/test/java/com/swp/testing/CalculatorTest.java
package com.swp.testing;

import org.junit.jupiter.api.*;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.*;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit тест без Spring контекст.
 * Тества Calculator директно – бърз, изолиран.
 */
class CalculatorTest {

    private Calculator calc;

    @BeforeEach
    void setUp() {
        calc = new Calculator();
    }

    @Test
    void testAdd() {
        assertEquals(5.0, calc.add(2, 3));
    }

    @Test
    void testSubtract() {
        assertEquals(1.0, calc.subtract(3, 2));
    }

    @Test
    void testMultiply() {
        assertEquals(6.0, calc.multiply(2, 3));
    }

    @Test
    void testDivide() {
        assertEquals(2.5, calc.divide(5, 2));
    }

    @Test
    void testDivideByZeroThrowsException() {
        IllegalArgumentException ex = assertThrows(
            IllegalArgumentException.class,
            () -> calc.divide(10, 0)
        );
        assertTrue(ex.getMessage().contains("нула"));
    }

    /** Параметризиран тест – изпълнява се 3 пъти */
    @ParameterizedTest
    @CsvSource({"1,2,3", "0,0,0", "-1,1,0"})
    void testAddParameterized(double a, double b, double expected) {
        assertEquals(expected, calc.add(a, b));
    }

    @Test
    void testAddIsCommutative() {
        assertEquals(calc.add(3, 7), calc.add(7, 3));
    }
}
